File: app/src/main/java/com/mx/gillustrated/adapter/DataListAdapter.java
package com.mx.gillustrated.adapter;

import java.io.File;
import java.util.List;

import com.mx.gillustrated.R;
import com.mx.gillustrated.activity.BaseActivity;
import com.mx.gillustrated.activity.MainActivity;
import com.mx.gillustrated.common.MConfig;
import com.mx.gillustrated.component.ResourceController;
import com.mx.gillustrated.util.CommonUtil;
import com.mx.gillustrated.vo.CardInfo;

import android.app.ActionBar;
import android.content.Context;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.TextView;

public class DataListAdapter extends BaseAdapter {

	private MainActivity mcontext;
	private LayoutInflater layoutInflator;
	private List<CardInfo> list;

	public DataListAdapter() {
	}

	public DataListAdapter(MainActivity context, List<CardInfo> items) {
		mcontext = context;
		layoutInflator = LayoutInflater.from(mcontext);
		list = items;
	}

	@Override
	public int getCount() {
		return list.size();
	}

	@Override
	public Object getItem(int arg0) {
		return list.get(arg0);
	}

	@Override
	public long getItemId(int arg0) {
		return arg0;
	}

	@Override
	public View getView(int arg0, View convertView, ViewGroup arg2) {
		Component component = null;

		if (convertView == null) {
			convertView = layoutInflator.inflate(
					R.layout.adapter_mainlist, null);
			component= new Component();
			component.ivHeader = (ImageView) convertView
					.findViewById(R.id.ivHeader);
			component.tvName = (TextView) convertView
					.findViewById(R.id.tvName);
			component.tvFrontName = (TextView) convertView
					.findViewById(R.id.tvFrontName);
			component.tvAttr = (TextView) convertView
					.findViewById(R.id.tvAttr);
			component.tvCost = (TextView) convertView
					.findViewById(R.id.tvCost);
			component.tvHP = (TextView) convertView
					.findViewById(R.id.tvHP);
			component.tvAttack = (TextView) convertView
					.findViewById(R.id.tvAttack);
			component.tvDefense = (TextView) convertView
					.findViewById(R.id.tvDefense);
			component.tvExtra1 = (TextView) convertView
					.findViewById(R.id.tvExtra1);
			component.tvExtra2 = (TextView) convertView
					.findViewById(R.id.tvExtra2);
			component.tvNid = (TextView) convertView
					.findViewById(R.id.tvNid);
			component.tvTotal = (TextView) convertView
					.findViewById(R.id.tvTotal);
			component.ivExtraGap1 = (ImageView) convertView
					.findViewById(R.id.ivExtra1Gap);
			component.ivExtraGap2 = (ImageView) convertView
					.findViewById(R.id.ivExtra2Gap);
			component.ivCostGap = (ImageView) convertView
					.findViewById(R.id.ivCostGap);
			convertView.setTag(component);

			if( list.get(arg0).getNid() > 0)
				component.tvNid.setVisibility(View.VISIBLE);
			else
				component.tvNid.setVisibility(View.GONE);

			ResourceController resourceController = new ResourceController(mcontext, list.get(arg0).getGameId());
			if( "E1".equals(resourceController.getNumber4())){
				component.ivExtraGap1.setVisibility(View.GONE);
				component.tvExtra1.setVisibility(View.GONE);
			}
			if( "E2".equals(resourceController.getNumber5())){
				component.ivExtraGap2.setVisibility(View.GONE);
				component.tvExtra2.setVisibility(View.GONE);
			}
			if(!mcontext.mSP.getBoolean(mcontext.SHARE_SHOW_COST_COLUMN + list.get(arg0).getGameId(), false)){
				component.ivCostGap.setVisibility(View.GONE);
				component.tvCost.setVisibility(View.GONE);
			}

		}
		else
		{
			component = (Component) convertView.getTag();
		}

		boolean hasHeader = "Y".equals(list.get(arg0).getProfile());
		boolean showHeader = mcontext.mSP.getBoolean(BaseActivity.SHARE_SHOW_HEADER_IMAGES + list.get(arg0).getGameId(), false);
		if(showHeader && hasHeader){
			component.tvTotal.setVisibility(View.GONE);
			component.ivHeader.setVisibility(View.VISIBLE);
			try {
				File imageDir = new File(Environment.getExternalStorageDirectory(), MConfig.SD_HEADER_PATH + "/" + list.get(arg0).getGameId());
				File file = new File(imageDir.getPath(), list.get(arg0).getId() + "_h.png");
				if(file.exists())
				{
					Bitmap bmp = MediaStore.Images.Media.getBitmap(mcontext.getContentResolver(), Uri.fromFile(file));
					component.ivHeader.setImageBitmap(bmp);
				}
				else
					component.ivHeader.setImageBitmap(null);

			} catch (Exception e) {
				e.printStackTrace();
			}
		}
		else if(showHeader) {
			component.tvTotal.setVisibility(View.GONE);
			component.ivHeader.setVisibility(View.GONE);

		}else {
			component.tvTotal.setVisibility(View.VISIBLE);
			component.ivHeader.setVisibility(View.GONE);
			if( !Double.isNaN(Double.valueOf(list.get(arg0).getMaxHP())) &&
					!Double.isNaN(Double.valueOf(list.get(arg0).getMaxAttack())) &&
					!Double.isNaN(Double.valueOf(list.get(arg0).getMaxDefense())) &&
					!Double.isNaN(Double.valueOf(list.get(arg0).getExtraValue1())) &&
					!Double.isNaN(Double.valueOf(list.get(arg0).getExtraValue2()))){
				int total = Integer.parseInt(list.get(arg0).getMaxHP()) +
						Integer.parseInt(list.get(arg0).getMaxAttack()) +
						Integer.parseInt(list.get(arg0).getMaxDefense()) +
						Integer.parseInt(list.get(arg0).getExtraValue1()) +
						Integer.parseInt(list.get(arg0).getExtraValue2());
				component.tvTotal.setText( String.valueOf(total));
			}


		}

		component.tvName.setText(list.get(arg0).getName());
		component.tvFrontName.setText(list.get(arg0).getFrontName());
		String attr = list.get(arg0).getAttr();
		component.tvAttr.setText(attr);
		component.tvCost.setText(String.valueOf(list.get(arg0).getCost()));
		component.tvHP.setText(String.valueOf(list.get(arg0).getMaxHP()));
		component.tvAttack.setText(String.valueOf(list.get(arg0).getMaxAttack()));
		component.tvDefense.setText(String.valueOf(list.get(arg0).getMaxDefense()));
		if(list.get(arg0).getExtraValue1() != null)
			component.tvExtra1.setText(String.valueOf(list.get(arg0).getExtraValue1()));
		else
			component.tvExtra1.setText("");
		if(list.get(arg0).getExtraValue2() != null)
			component.tvExtra2.setText(String.valueOf(list.get(arg0).getExtraValue2()));
		else
			component.tvExtra2.setText("");
		component.tvNid.setText(String.valueOf(list.get(arg0).getNid()));

		return convertView;
	}

	private static class Component{
		 public ImageView ivHeader;
		 public TextView tvFrontName;
		 public TextView tvName;
		 public TextView tvAttr;
		 public TextView tvCost;
		 public TextView tvHP;
		 public TextView tvAttack;
		 public TextView tvDefense;
		 public TextView tvExtra1;
		 public TextView tvExtra2;
		 public TextView tvNid;
		 public TextView tvTotal;
		 public ImageView ivExtraGap1;
		 public ImageView ivExtraGap2;
		 public ImageView ivCostGap;
	}
}

File: app/src/main/java/com/mx/gillustrated/adapter/DataListAdapter.kt
package com.mx.gillustrated.adapter

import java.io.File
import com.mx.gillustrated.activity.BaseActivity
import com.mx.gillustrated.activity.MainActivity
import com.mx.gillustrated.common.MConfig
import com.mx.gillustrated.component.ResourceController
import com.mx.gillustrated.util.CommonUtil
import com.mx.gillustrated.vo.CardInfo
import android.net.Uri
import android.os.Environment
import android.provider.MediaStore
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.mx.gillustrated.databinding.AdapterMainlistBinding

class DataListAdapter constructor(private val mContext: MainActivity, private val list: List<CardInfo>) : BaseAdapter() {

    private val layoutInflater: LayoutInflater = LayoutInflater.from(mContext)

    override fun getCount(): Int {
        return list.size
    }

    override fun getItem(arg0: Int): Any {
        return list[arg0]
    }

    override fun getItemId(arg0: Int): Long {
        return arg0.toLong()
    }

    override fun getView(arg0: Int, convertViews: View?, arg2: ViewGroup): View {
        var convertView = convertViews
        lateinit var component: AdapterMainlistBinding

        if (convertView == null) {
            component = AdapterMainlistBinding.inflate(layoutInflater, arg2, false)
            convertView = component.root
            convertView.tag = component

            val resourceController = ResourceController(mContext, list[arg0].gameId)
            if ("E1" == resourceController.number4) {
                component.ivExtra1Gap.visibility = View.GONE
                component.tvExtra1.visibility = View.GONE
            }
            if ("E2" == resourceController.number5) {
                component.ivExtra2Gap.visibility = View.GONE
                component.tvExtra2.visibility = View.GONE
            }
            if (!mContext.mSP.getBoolean(BaseActivity.SHARE_SHOW_COST_COLUMN + list[arg0].gameId, false)) {
                component.ivCostGap.visibility = View.GONE
                component.tvCost.visibility = View.GONE
            }
        } else {
            component = convertView.tag as AdapterMainlistBinding
        }

        val hasHeader = "Y" == list[arg0].profile
        val showHeader = mContext.mSP.getBoolean(BaseActivity.SHARE_SHOW_HEADER_IMAGES + list[arg0].gameId, false)
        if (showHeader && hasHeader) {
            component.tvTotal.visibility = View.GONE
            component.ivHeader.visibility = View.VISIBLE
            try {
                val imageDir = File(Environment.getExternalStorageDirectory(), MConfig.SD_HEADER_PATH + "/" + list[arg0].gameId)
                val file = File(imageDir.path, list[arg0].id.toString() + "_h.png")
                if (file.exists()) {
                    val bmp = MediaStore.Images.Media.getBitmap(mContext.contentResolver, Uri.fromFile(file))
                    component.ivHeader.setImageBitmap(bmp)
                } else
                    component.ivHeader.setImageBitmap(null)

            } catch (e: Exception) {
                e.printStackTrace()
            }

        } else if (showHeader) {
            component.tvTotal.visibility = View.GONE
            component.ivHeader.visibility = View.GONE

        } else {
            component.tvTotal.visibility = View.VISIBLE
            component.ivHeader.visibility = View.GONE
            if ("" != list[arg0].maxHP &&
                    "" != list[arg0].maxAttack &&
                    "" != list[arg0].maxDefense &&
                    "" != list[arg0].extraValue1 &&
                    "" != list[arg0].extraValue2 &&
                    CommonUtil.isNumeric2(list[arg0].maxHP) &&
                    CommonUtil.isNumeric2(list[arg0].maxAttack) &&
                    CommonUtil.isNumeric2(list[arg0].maxDefense) &&
                    CommonUtil.isNumeric2(list[arg0].extraValue1) &&
                    CommonUtil.isNumeric2(list[arg0].extraValue2)) {
                val total = Integer.parseInt(list[arg0].maxHP!!) +
                        Integer.parseInt(list[arg0].maxAttack!!) +
                        Integer.parseInt(list[arg0].maxDefense!!) +
                        Integer.parseInt(list[arg0].extraValue1!!) +
                        Integer.parseInt(list[arg0].extraValue2!!)
                component.tvTotal.text = total.toString()
            }


        }

        component.tvName.text = list[arg0].name
        component.tvFrontName.text = list[arg0].frontName
        val attr = list[arg0].attr
        component.tvAttr.text = attr
        component.tvCost.text = list[arg0].cost.toString()
        component.tvHP.text = list[arg0].maxHP.toString()
        component.tvAttack.text = list[arg0].maxAttack.toString()
        component.tvDefense.text = list[arg0].maxDefense.toString()
        if (list[arg0].extraValue1 != null)
            component.tvExtra1.text = list[arg0].extraValue1.toString()
        else
            component.tvExtra1.text = ""
        if (list[arg0].extraValue2 != null)
            component.tvExtra2.text = list[arg0].extraValue2.toString()
        else
            component.tvExtra2.text = ""

        component.tvNid.text = list[arg0].nid.toString()
        if (list[arg0].nid > 0)
            component.tvNid.visibility = View.VISIBLE
        else
            component.tvNid.visibility = View.GONE

        return convertView
    }


}
